src/cascadia/WindowsTerminal/WindowEmperor.cpp

```diff
@@ -18,600 +24,1015 @@ using namespace ::Microsoft::Console;
+        CreateNewWindow(std::move(request));
+    }
+}
 
-    // Set the number of open windows (so we know if we are the last window)
-    // and subscribe for updates if there are any changes to that number.
+// This is an implementation-detail of _dispatchCommandline().
+safe_void_coroutine WindowEmperor::_dispatchCommandlineCurrentDesktop(winrt::TerminalApp::CommandlineArgs args)
+{
+    std::weak_ptr<AppHost> mostRecentWeak;
 
-    _revokers.WindowCreated = _manager.WindowCreated(winrt::auto_revoke, { this, &WindowEmperor::_numberOfWindowsChanged });
-    _revokers.WindowClosed = _manager.WindowClosed(winrt::auto_revoke, { this, &WindowEmperor::_numberOfWindowsChanged });
+    if (winrt::guid currentDesktop; VirtualDesktopUtils::GetCurrentVirtualDesktopId(reinterpret_cast<GUID*>(&currentDesktop)))
+    {
+        int64_t max = INT64_MIN;
+        for (const auto& w : _windows)
+        {
+            const auto lastActivatedTime = w->GetLastActivatedTime();
+            const auto desktopId = co_await w->GetVirtualDesktopId();
+            if (desktopId == currentDesktop && lastActivatedTime > max)
+            {
+                max = lastActivatedTime;
+                mostRecentWeak = w;
+            }
+        }
+    }
 
-    // If a previous session of Windows Terminal stored buffer_*.txt files, then we need to clean all those up on exit
-    // that aren't needed anymore, even if the user disabled the ShouldUsePersistedLayout() setting in the meantime.
+    // GetVirtualDesktopId(), as current implemented, should always return on the main thread.
+    _assertIsMainThread();
+
+    const auto mostRecent = mostRecentWeak.lock();
+    auto window = mostRecent.get();
+
+    if (!window)
     {
-        const auto state = ApplicationState::SharedInstance();
-        const auto layouts = state.PersistedWindowLayouts();
-        _requiresPersistenceCleanupOnExit = layouts && layouts.Size() > 0;
+        window = _mostRecentWindow();
+    }
+
+    if (window)
+    {
+        window->DispatchCommandline(std::move(args));
+    }
+    else
+    {
+        CreateNewWindow(winrt::TerminalApp::WindowRequestedArgs{ 0, std::move(args) });
     }
 }
 
-// sender and args are always nullptr
-void WindowEmperor::_numberOfWindowsChanged(const winrt::Windows::Foundation::IInspectable&,
-                                            const winrt::Windows::Foundation::IInspectable&)
+bool WindowEmperor::_summonWindow(const SummonWindowSelectionArgs& args) const
 {
-    // If we closed out the quake window, and don't otherwise need the tray
-    // icon, let's get rid of it.
-    _checkWindowsForNotificationIcon();
+    AppHost* window = nullptr;
+
+    if (args.WindowID)
+    {
+        for (const auto& w : _windows)
+        {
+            if (w->Logic().WindowProperties().WindowId() == args.WindowID)
+            {
+                window = w.get();
+                break;
+            }
+        }
+    }
+    else if (!args.WindowName.empty())
+    {

```

Comment:
📝: you know, the code is cleaner if it's one `for` loop with the `if(args.WindowID) / else if (!args.WindowName.empty())` statement on the inside, but it's less efficient then. :shrug:

Reply:
Nah, it'd would not really be a big difference (branch prediction + it usually never loops >10 times anyway). This is also not really intentional, and I'll just move the if condition inside.

Reply:
Eh, you know what, I'll leave this as-is for now, because it makes the `window = _mostRecentWindow();` branch at the end easier to express. If anyone ever happens to change this code I would not mind at all.